Make aot_module_simplified accept fake tensors

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89773
* #89672
* #89671
* __->__ #89670
* #89669
* #89736

Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make aot_module_simplified accept fake tensors

Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

voznesenskym commented on functorch/_src/aot_autograd.py:
```diff
@@ -1488,18 +1488,30 @@ def create_aot_dispatcher_function(
+    # Check flat_args to see if they're already fake.  If so, use that fake
+    # mode instead.
+
+    for x in flat_args:
+        if isinstance(x, FakeTensor):
+            fake_mode = x.fake_mode
+            break
```
I have a util I wrote for this, in the parent PR

```
def fake_mode_from_tensors(inputs: List[Any]):
    """
    Takes a list of anything, unflattened is fine, returns a fake_mode
    if any are fake. All fake modes on all fake tensors must be identical.
    Returns None if no fake_mode is fine
    """
    flat_inputs, _ = tree_flatten(inputs)
    fake_mode = None
    for flat_input in flat_inputs:
        if isinstance(flat_input, torch._subclasses.FakeTensor):
            if fake_mode is None:
                fake_mode = flat_input.fake_mode
            else:
                assert fake_mode == flat_input.fake_mode
    return fake_mode
```

Feel free to replace the continue-and-check-all-same-fake-mode assert w/ a break as you have it, but I do thing this should be in utils

ezyang replied:
I didn't use this utility because `flat_args` is guaranteed to be a flat list of tensors here, whereas the utility does a flatten first. Better not to use pytree if you don't need it.

voznesenskym commented on functorch/_src/aot_autograd.py:
```diff
@@ -1488,18 +1488,30 @@ def create_aot_dispatcher_function(
 
         def process_inputs(flat_args):
-            if config.use_fake_tensor:
+            if config.use_fake_tensor or isinstance(fake_mode, FakeTensorMode):
                 def convert(idx, x):
                     if not isinstance(x, torch.Tensor):
                         return x
+                    if isinstance(x, FakeTensor):
+                        assert x.fake_mode is fake_mode
+                        return x
                     if idx < aot_config.num_params_buffers and config.static_weight_shapes:
                         return fake_mode.from_tensor(x, static_shapes=True)
                     return fake_mode.from_tensor(x, static_shapes=False)
```
Ive gone back and forth on the signal here.

1) Dynamo sets `config.use_fake_tensor` correctly
2) We check `config.use_fake_tensor` in other place
3) You should never be in a state where you are in `config.use_fake_tensor == False` and getting a fake mode

So maybe instead we either:

A) Assert that we are in fake config mode if we see a fake tensor in inputs on L1495 and use the presence of fake_mode to be analogous to config.use_fake_tensor (since you can only get one via inputs in config.use_fake_tensor or by making your own on L1500, also within config)

B) Or drop the isinstance(fake_mode) checks and assume that if we have the config set, we have the fake mode, and guard that with an assert.

ezyang replied:
I don't think this is worth litigating much, because I think we should delete this config option (e.g., https://github.com/pytorch/pytorch/pull/89663 ).

ezyang committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

voznesenskym committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

voznesenskym committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make aot_module_simplified accept fake tensors"


Strategy taken from voz's #89392 but my implementation strategy
is a bit different.

If a fake tensor is provided, we use its FakeTensorMode
(and more importantly, its ShapeEnv--this is what is tested
in the new unit test).  Only one tensor needs to be fake;
if nothing is fake we just make a fresh mode as before.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee reviewed (commented):
Is this just a temporary PR that will no longer be needed once we do fakification at the Dynamo level? (or at least prior to `create_aot_dispatcher_function`?). 

I thought the contract with `create_aot_dispatcher_function` was going to be that it was passed a list of fake tensors.
